doc/mgmt/SONiC_DNS_Support.md

```diff
@@ -0,0 +1,433 @@
+```
+### 3.6.4 Click command support
+#### 3.6.4.1 Configuration support
+Below click commands are supported for DNS configurations
+oot@sonic:~# config dns
+Usage: config dns [OPTIONS] COMMAND [ARGS]...
+
+  DNS command line
+
+Options:
+  -?, -h, --help  Show this message and exit.
+
+Commands:
+  add          Specify a DNS name server
+  delete       Delete a DNS name server
+  source_intf  DNS source interface configurations
+root@sonic:~#
+root@sonic:~# config dns source_intf
+Usage: config dns source_intf [OPTIONS] COMMAND [ARGS]...
+
+  DNS source interface configurations
+
+Options:
+  -?, -h, --help  Show this message and exit.
+
+Commands:
+  add     DNS source interface add configuration
+  delete  DNS source interface delete configuration
+root@sonic:~#
+```
+root@sonic:~# ping www.yahoo.com
+ping: www.yahoo.com: Temporary failure in name resolution
+
+root@sonic:~# config dns add 10.11.0.1
+
+root@sonic:~# ping www.yahoo.com
+PING new-fp-shed.wg1.b.yahoo.com (98.137.246.8) 56(84) bytes of data.
+64 bytes from media-router-fp2.prod1.media.vip.gq1.yahoo.com (98.137.246.8): icmp_seq=1 ttl=52 time=24.2 ms
+64 bytes from media-router-fp2.prod1.media.vip.gq1.yahoo.com (98.137.246.8): icmp_seq=2 ttl=52 time=23.9 ms
+64 bytes from media-router-fp2.prod1.media.vip.gq1.yahoo.com (98.137.246.8): icmp_seq=3 ttl=52 time=23.9 ms
+^C
+--- new-fp-shed.wg1.b.yahoo.com ping statistics ---
+3 packets transmitted, 3 received, 0% packet loss, time 2001ms
+rtt min/avg/max/mdev = 23.910/24.033/24.262/0.241 ms
+root@sonic:~#
+```
+#### 3.6.4.2 Show support
+Below Click command is used to dump the DNS configurations
+```
+root@sonic:~# show hosts
+Source Interface : Loopback0
+Name servers are : 10.11.0.1, 2001:aa:aa::a
+root@sonic:~#
+```
+# 4 Flow Diagrams
+
+# 5 Error Handling
+
+# 6 Serviceability and Debug
+
+# 7 Warm Boot Support
+This support is added in the hostcfgd and hence no explicit handling is needed.
+
+# 8 Scalability
+
+# 9 Unit Test
+
+The unit-test for this feature will include:
+#### Configuration via CLI
+
+| Test Name | Test Description |
+| :-------- | :----- |
+| Configure DNS server | Verify DNS servers are present in the DNS_SERVER table (configDB) and the same is reflected in the resolv.conf file |
+| Delete DNS server | Verify DNS servers are not present in the DNS_SERVER table (configDB) and the same is reflected in the resolv.conf file  |
+| Verify max DNS servers | Verify more than 6 servers are not allowed |
+| Configure DNS source interface | Verify source interface is present in the DNS table (configDB) and DNS query packets are transmitted with the source IP present in source interface |
+| Delete DNS source interface | Verify source interface is not present in the DNS table (configDB) and DNS query packets are not transmitted with the source IP present in source interface |
+| Configure mgmt VRF | Verify that DNS query packets (from mgmt VRF) are transmitted based on nameservers present in resolv.conf file |

```

Comment:
When the mgmt-VRF is enabled, should not all DNS queries be resolved using Management VRF only? What is the meaning of configure mgmt-VRF ? How does the application tell the DNS to use Management VRF ?

Reply:
whichever IP comes first from the table based on source-intf match will be chosen.

Reply:
Looks like this was a reply to the previous question. So the question about Mgmt VRF still stands. Is the Management VRF being supported for DNS queries?  Seems like there was some review comments earlier about Management VRF not being supported earlier?

Reply:
Yes, DNS on mgmt VRF is supported, no special handling is required, any traffic originating from mgmt VRF refers the "/etc/resolv.conf" for name resolution.

Reply:
So, is the "DNS on mgmt VRF" referring to which VRF the DNS server should be looked up in? Or is it referring to listening for local client DNS queries on the loopback interface of a particular VRF ? ("lo" -- default VRF vs "lo-m" -- Mgmt VRF ?) It is confusing what is the intent of the mgmt-VRF support for DNS. Can you please clarify?

Reply:
Irrespective of where (default VRF or mgmt VRF) we originate the request (e.g ping with domain name), name resolution refers the file "/etc/resolv.conf" file to get the name server IP to send the DNS queries, there seems to be no service running for name resolution, the name resolution is expected to happen by default with POSIX APIs.

Reply:
Are the POSIX apis VRF aware ? If they are not VRF aware, then we cannot claim Management VRF support. It looks like DNS is not supported when Management VRF is enabled. Can you confirm this point ? Unless the application which is making the POSIX api has somehow constrained the api (outside of the api) to use Management VRF, it will not be able to use Management VRF. Is this reasoning correct?

Reply:
We tested the name resolution in mgmt VRF for the ping. please see below for more information, are you referring to something else?
root@sonic:~# cgexec -g l3mdev:mgmt bash
root@sonic:~# ping www.yahoo.com
PING new-fp-shed.wg1.b.yahoo.com (98.138.219.232) 56(84) bytes of data.
64 bytes from media-router-fp2.prod1.media.vip.ne1.yahoo.com (98.138.219.232): icmp_seq=1 ttl=52 time=39.5 ms
64 bytes from media-router-fp2.prod1.media.vip.ne1.yahoo.com (98.138.219.232): icmp_seq=2 ttl=52 time=39.2 ms
^C
--- new-fp-shed.wg1.b.yahoo.com ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 39.258/39.396/39.535/0.242 ms
root@sonic:~# exit
exit
oot@sonic:~# cgexec -g l3mdev:mgmt ping www.google.com
PING www.google.com (172.217.164.100) 56(84) bytes of data.
64 bytes from sfo03s18-in-f4.1e100.net (172.217.164.100): icmp_seq=1 ttl=53 time=3.06 ms
64 bytes from sfo03s18-in-f4.1e100.net (172.217.164.100): icmp_seq=2 ttl=53 time=2.69 ms
^C
--- www.google.com ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1000ms
rtt min/avg/max/mdev = 2.696/2.881/3.067/0.193 ms
root@sonic:~#

root@sonic:~# cat /etc/resolv.conf
###############################################################################
#
#  /etc/resolv.conf    Configuration file for DNS nameserver.

#Set nameserver
nameserver 8.8.8.8

root@sonic:~#

Reply:
Yes, the application is constrained using cgexec. This works only for some applications. Other applications like "docker pull" talk to another daemon -- dockerd. So they would not work in those cases. Yes, cgexec is limiting the application scope (in this case ping scope) to use l3mdev:mgmt -- Management VRF

Reply:
docker pull doesnt work well with current mgmt VRF implementation with cgcreate, if we go with the workaround we have to allow default table lookup as a last resort from default VRF, DNS loookup will work fine in that case as well, hope we have covered all cases.

```

root@sonic:~# cgexec -g l3mdev:mgmt ping www.google.com
PING www.google.com (172.217.164.100) 56(84) bytes of data.
64 bytes from sfo03s18-in-f4.1e100.net (172.217.164.100): icmp_seq=1 ttl=53 time=2.97 ms
64 bytes from sfo03s18-in-f4.1e100.net (172.217.164.100): icmp_seq=2 ttl=53 time=2.66 ms
^C
--- www.google.com ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 2.667/2.819/2.972/0.161 ms
root@sonic:~# ping www.google.com
PING www.google.com (172.217.164.100) 56(84) bytes of data.
64 bytes from sfo03s18-in-f4.1e100.net (172.217.164.100): icmp_seq=1 ttl=53 time=3.05 ms
64 bytes from sfo03s18-in-f4.1e100.net (172.217.164.100): icmp_seq=2 ttl=53 time=2.67 ms
64 bytes from sfo03s18-in-f4.1e100.net (172.217.164.100): icmp_seq=3 ttl=53 time=2.68 ms
^C
--- www.google.com ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2002ms
rtt min/avg/max/mdev = 2.675/2.804/3.053/0.181 ms
root@sonic:~#
```